[doc] Document usage of Mono.zip() along with Mono<Void> (#3306) (#3514)
reactor/reactor-core: docs/asciidoc/faq.adoc

----
====

[[faq.monoZipEmptyCompletion]]
== Using `zip` along with empty-completed publishers

When using the `zip` operator along with empty-completed publishers, it is important to be aware of the following behavior.

Consider the following test case:

====
[source,java]
----
    @Test
    public void testZipEmptyCompletionAllSubscribed() {
        AtomicInteger cnt = new AtomicInteger();
        Mono<Integer> mono1 = Mono.create(sink -> {
            cnt.incrementAndGet();
            sink.success();
        });
        Mono<Integer> mono2 = Mono.create(sink -> {
            cnt.incrementAndGet();
            sink.success();
        });
        Mono<Integer> zippedMono = Mono.zip(mono1, mono2, (v1, v2) -> v1);
        zippedMono.subscribe();
        assertEquals(2, cnt.get());
    }
----
====

> While in this case the resulting `zippedMono` does not complete immediately upon empty completion of `mono1`, such behaviour is not guaranteed for all cases. For instance, consider the following test case:

Reviewer: ```suggestion
While in this case the resulting `zippedMono` subscribes to both `mono1` and `mono2`, such behaviour is not guaranteed for all cases. For instance, consider the following test case:
```

Reviewer: I think the language mentioning completing immediately is not the essence here and can be confusing. Whether it completes before or after is an implementation detail, but the user is most probably interested in the act of subscribing and processing the sources.

====
[source,java]
----
    @Test
    public void testZipEmptyCompletionOneSubscribed() {
        AtomicInteger cnt = new AtomicInteger();
        Mono<Integer> mono1 = Mono.create(sink -> {
            cnt.incrementAndGet();
            sink.success();
        });
        Mono<Integer> mono2 = Mono.create(sink -> {
            cnt.incrementAndGet();
            sink.success();
        });
        Mono<Integer> mono3 = Mono.create(sink -> {
            cnt.incrementAndGet();
            sink.success();
        });
        Mono<Integer> zippedMono = Mono.zip(mono1, Mono.zip(mono2, mono3, (v1, v2) -> v1), (v1, v2) -> v1);
        zippedMono.subscribe();
        assertEquals(1, cnt.get());
    }
----
====

> In this case upon empty completion of `mono1`, `zippedMono` completes immediately.

Reviewer: ```suggestion
In this case upon empty completion of `mono1`, `zippedMono` completes immediately and does not subscribe to `mono2` and `mono3`.
```

> In cases where `zip` operator is used to combine empty-completed publishers, it is therefore not guaranteed that the resulting publisher will complete only upon completion of all publishers to be combined. In other words, while in some cases this behaviour might be different, it is not guaranteed that the resulting publisher will not complete immediately upon empty completion of the first publisher to be zipped.

Reviewer: Please consider rewriting this according to the above suggestions around subscription and not completion.
